core/src/main/scala/schema/provider/MetadataSchemaProvider.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2017 47 Degrees, LLC. <http://www.47deg.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package freestyle.cassandra
+package schema.provider
+
+import cats.implicits._
+import cats.~>
+import com.datastax.driver.core._
+import freestyle._
+import freestyle.FreeS
+import freestyle.cassandra.schema.provider.metadata.SchemaConversions
+import freestyle.cassandra.schema.{SchemaDefinition, SchemaDefinitionProviderError}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
+import scala.language.postfixOps
+
+class MetadataSchemaProvider(cluster: Cluster)
+    extends SchemaDefinitionProvider
+    with SchemaConversions {
+
+  def extractTables(keyspaceMetadata: KeyspaceMetadata): List[AbstractTableMetadata] =
+    keyspaceMetadata.getTables.asScala.toList
+
+  def extractIndexes(tableMetadataList: List[AbstractTableMetadata]): List[IndexMetadata] =
+    tableMetadataList.flatMap {
+      case (t: TableMetadata) => t.getIndexes.asScala.toList
+      case _                  => Nil
+    }
+
+  def extractUserTypes(keyspaceMetadata: KeyspaceMetadata): List[UserType] =
+    keyspaceMetadata.getUserTypes.asScala.toList
+
+  override def schemaDefinition: Either[SchemaDefinitionProviderError, SchemaDefinition] = {
+
+    import scala.concurrent.ExecutionContext.Implicits.global
+
+    import freestyle.async.implicits._
+    import freestyle.cassandra.api._
+    import freestyle.cassandra.handlers.implicits._
+
+    implicit val clusterAPIInterpreter: ClusterAPI.Op ~> Future =
+      clusterAPIHandler[Future] andThen apiInterpreter[Future, Cluster](cluster)
+
+    def metadataF[F[_]](implicit clusterAPI: ClusterAPI[F]): FreeS[F, Metadata] =
+      for {
+        _        <- clusterAPI.connect
+        metadata <- clusterAPI.metadata
+        _        <- clusterAPI.close
+      } yield metadata
+
+    Either.catchNonFatal {
+      Await.result(metadataF[ClusterAPI.Op].interpret[Future], 10.seconds)
```

Comment:
I can see that `schemaDefinition` needs to return an `Either[...]` at the moment, I am not sure if we can change that somehow.

Currently it is unfortunate twice: first we immediately have to interpret a `FreeS` and then we we have to block on a `Future`. It doesn't appear to be trivial to change this though ... 

If there is no way we can do that, I would put the `Await` on the end.

```scala
val fut: Future[SchemaDefinitionProviderError, SchemaDefinition] =
  metadataF[ClusterAPI.Op].interpret[Future].attempt.map {
    _.leftMap (SchemaDefinitionProviderError(_)) flatMap { metadata =>
      // ...
    }
  }
Await.result(fut, 10.seconds) // :(
```